standard/statements.md

```diff
@@ -1029,12 +1033,16 @@ The `foreach` statement enumerates the elements of a collection, executing an em
 
 ```ANTLR
 foreach_statement
-    : 'foreach' '(' local_variable_type identifier 'in' expression ')'
-      embedded_statement
+    : 'foreach' '(' ('ref' 'readonly'?)? local_variable_type identifier 'in' 
+      expression ')' embedded_statement
     ;
 ```
 
-The *local_variable_type* and *identifier* of a `foreach` statement declare the ***iteration variable*** of the statement. If the `var` identifier is given as the *local_variable_type*, and no type named var is in scope, the iteration variable is said to be an ***implicitly typed iteration variable***, and its type is taken to be the iteration type of the `foreach` statement, as specified below. The iteration variable corresponds to a read-only local variable with a scope that extends over the embedded statement. During execution of a `foreach` statement, the iteration variable represents the collection element for which an iteration is currently being performed. A compile-time error occurs if the embedded statement attempts to modify the iteration variable (via assignment or the `++` and `--` operators) or pass the iteration variable as a `ref` or `out` parameter.
+The *local_variable_type* and *identifier* of a foreach statement declare the ***iteration variable*** of the statement. If the `var` identifier is given as the *local_variable_type*, and no type named `var` is in scope, the iteration variable is said to be an ***implicitly typed iteration variable***, and its type is taken to be the element type of the `foreach` statement, as specified below.
+
+If the *foreach_statement* contains both or neither `ref` and `readonly`, the iteration variable denotes a variable that is treated as read-only. Otherwise, if *foreach_statement* contains `ref` without `readonly`, the iteration variable denotes a variable that shall be writable.
+
+The iteration variable corresponds to a local variable with a scope that extends over the embedded statement. During execution of a `foreach` statement, the iteration variable represents the collection element for which an iteration is currently being performed. If the iteration variable denotes a read-only variable, a compile-time error occurs if the embedded statement attempts to modify it (via assignment or the `++` and `--` operators) or pass it as a `ref` or `out` parameter.
```

Comment:
The previous paragraph said it is readonly, and this paragraph says that is optional.

I suspect the constraint is that you cannot use the `= ref` operator on the variable.

Reply:
I think the text is correct, if awkward. I'm making the following suggestion to simplify. The previous paragraph states the conditions under which the variable is a read-only variable. If it's read-only, it can't be modified (by the definition of read-only). So, that sentence is just restating the obvious:

```suggestion
The iteration variable corresponds to a local variable with a scope that extends over the embedded statement. During execution of a `foreach` statement, the iteration variable represents the collection element for which an iteration is currently being performed.
```

Reply:
IIn what version of C# is a `ref` foreach iteration variable supported?

Reply:
Added in 7.3.  Here's an example:

```csharp
    public class C2
    {
        private int[] storage = new int[20];

        public void Initialize()
        {
            int index = 0;
            foreach (ref int i in storage.AsSpan())
            {
                i = index++;
            }
        }
    }
```

Reply:
Thanks @BillWagner, I'd failed to pin it down.

I think there is a big chunk missing from the `foreach` description – while it defines that the iteration variable maybe a reference variable it doesn’t describe that the type of the *expression* in this case must be one with a ref-returning enumerator. As your example shows arrays do not fit this requirement, you need to create a `Span` which does have such an enumerator.

At a quick read I’m not immediately confident that a ref-foreach will pass the compile-time processing description.

We need to review the description and add at least one example of a ref-foreach.

Reply:
Bill has created #804 to cover this. It definitely needs to be done before C# 7 though.